### PR TITLE
adapter-auto: defer adapter installation until required

### PR DESCRIPTION
**Pull request by gtm-nayan**
### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [ ] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [ ] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. All changesets should be `patch` until SvelteKit 1.0

### CODE REVIEW

**Review by dummdidumm — APPROVED** [review]
Makes sense to me - leaving this open so others can give their opinions on this

**Comment by Rich-Harris** [comment]
If we add new capabilities to adapters in future — for example, if they're allowed to control aspects of the dev environment — then this won't work

**Comment by Conduitry** [comment]
Probably we don't want the auto adapter to be affecting dev mode? It wouldn't have any of the env vars it looks at anyway. Presumably people should explicitly install their target adapter if they want to have it impact dev mode too.

**Comment by gtm-nayan** [comment]
To provide platform specific features in dev mode you'd have to know what platform you'd be deploying to, and at that point adapter-auto can be swapped out manually.

I thought the idea was that adapter-auto would serve as a minimal deployable setup for new users unfamiliar with `adapters` and users requiring more features can use specific ones for their platform. There's a warning in adapter-auto which reflects that sentiment: `If you plan on staying on this deployment platform, consider replacing @sveltejs/adapter-auto with ${match.module}. This will give you faster and more robust installs, and more control over deployment configuration.`

**Comment by Rich-Harris** [comment]
That's a fair point! I rescind my earlier comments. That said, I'm not 100% sure I understand what problem this solves?

File: packages/adapter-auto/index.js [hunk 1]
```diff
@@ -46,72 +43,78 @@ async function import_from_cwd(name) {
...
-					);
-				}
-			} else {
-				throw error;
+/** @typedef {import('@sveltejs/kit').Adapter} Adapter */
+
+/**
+ * @returns {Promise<Adapter | undefined>} The corresponding adapter for the current environment if found otherwise undefined
+ */
+async function get_adapter() {
```
Comment (benmccann) [inline packages/adapter-auto/index.js:51]: I'd make it clearer from the name that this is installing the adapter as `get` sounds side-effect free
```suggestion
async function install_adapter() {
```

**Comment by benmccann** [comment]
> I'm not 100% sure I understand what problem this solves?

Right now it's installing the adapter up-front, which seems to be doing the install in the middle of the Vite bundling causing two versions of SvelteKit to end up in the final bundle. By waiting until the bundle is written, hopefully that's fixed.

I'm not 100% sure how to test this other than merging it and updating the sites repo. There's probably a way, but I'd be just as happy to merge this even if it doesn't work as I think it's a nice way to structure the code, so I'd say we merge this and see if it works

File: packages/adapter-auto/index.js (hunk 1 above)
Comment (Rich-Harris) [reply packages/adapter-auto/index.js:51]: In some cases it _won't_ install anything though. We could make it `get_or_install_adapter` or whatever but i think it's clear enough from context

**Pull request merged**